README.md
# mms2r

  https://github.com/monde/mms2r

## DESCRIPTION

MMS2R by Mike Mondragon
https://github.com/monde/mms2r
https://rubygems.org/gems/mms2r
http://peepcode.com/products/mms2r-pdf

MMS2R is a library that decodes the parts of a MMS message to disk while
stripping out advertising injected by the mobile carriers.  MMS messages are
multipart email and the carriers often inject branding into these messages.  Use
MMS2R if you want to get at the real user generated content from a MMS without
having to deal with the cruft from the carriers.

If MMS2R is not aware of a particular carrier no extra processing is done to the
MMS other than decoding and consolidating its media.

MMS2R can be used to process any multipart email to conveniently access the
parts the mail is comprised of.

Contact the author to add additional carriers to be processed by the library.
Suggestions and patches appreciated and welcomed!

[![tip for next commit](http://tip4commit.com/projects/198.svg)](http://tip4commit.com/projects/198)

### Corpus of carriers currently processed by MMS2R:

* 1nbox/Idea: 1nbox.net
* 3 Ireland: mms.3ireland.ie
* Alltel: mms.alltel.com
* AT&T/Cingular/Legacy: mms.att.net, mm.att.net, txt.att.net, mmode.com, mms.mycingular.com, cingularme.com, mobile.mycingular.com pics.cingularme.com
* Bell Canada: txt.bell.ca
* Bell South / Suncom / SBC Global: bellsouth.net, sbcglobal.net
* Cricket Wireless: mms.mycricket.com
* Dobson/Cellular One: mms.dobson.net
* Helio: mms.myhelio.com
* Hutchison 3G UK Ltd: mms.three.co.uk
* INDOSAT M2: mobile.indosat.net.id
* LUXGSM S.A.: mms.luxgsm.lu
* Maroc Telecom / mms.mobileiam.ma
* MTM South Africa: mms.mtn.co.za
* NetCom (Norway): mms.netcom.no
* Nextel: messaging.nextel.com
* O2 Germany: mms.o2online.de
* O2 UK: mediamessaging.o2.co.uk
* Orange & Regional Oranges: orangemms.net, mmsemail.orange.pl, orange.fr
* PLSPICTURES.COM mms hosting: waw.plspictures.com
* PXT New Zealand: pxt.vodafone.net.nz
* Rogers of Canada: rci.rogers.com
* SaskTel: sasktel.com, sms.sasktel.com, cdma.sasktel.com
* Sprint: pm.sprint.com, messaging.sprintpcs.com, sprintpcs.com
* T-Mobile: tmomail.net, mmsreply.t-mobile.co.uk, tmo.blackberry.net
* TELUS Corporation (Canada): mms.telusmobility.com, msg.telus.com
* U.S. Cellular: mms.uscc.net
* UAE MMS: mms.ae
* Unicel: unicel.com, info2go.com (note: mobile number is tucked away in a text/plain part for unicel.com)
* Verizon: vzwpix.com, vtext.com, labwig.net
* Virgin Mobile: vmpix.com, pixmbl.com, vmobl.com, yrmobl.com
* Virgin Mobile of Canada: vmobile.ca
* Vodacom: mms.vodacom4me.co.za
* Cincinnati Bell: mms.gocbw.com

### Corpus of smart phones known to MMS2R:
* Apple iPhone variants
* Blackberry / Research In Motion variants
* Casio variants
* Droid variants
* Google / Nexus One variants
* HTC variants (T-Mobile Dash, Sprint HERO)
* LG Electronics variants
* Motorola variants
* Pantech variants
* Qualcom variants
* Samsung variants
* Sprint variants
* UTStarCom variants

### As Seen On The Internets - sites known to be using MMS2R in some fashion

* twitpic.com[http://www.twitpic.com/]
* Simplton[http://simplton.com/]
* fanchatter.com[http://www.fanchatter.com/]
* camura.com[http://www.camura.com/]
* eachday.com[http://www.eachday.com/]
* beenup2.com[http://www.beenup2.com/]
* snapmylife.com[http://www.snapmylife.com/]
* email the author to be listed here


## FEATURES

* #default_media and #default_text methods return a File that can be used in attachment_fu or Paperclip
* #process supports blocks for enumerating over the content of the MMS
* #process can be made lazy when :process => :lazy is passed to new
* logging is enabled when :logger => your_logger is passed to new
* an mms instance acts like a Mail object, any methods not defined on the instance are delegated to it's underlying Mail object
* #device_type? returns a symbol representing a device or smartphone type.  Known smartphones thus far: iPhone, BlackBerry, T-Mobile Dash, Droid, Samsung

## BOOKS

MMS2R, Making email useful
http://peepcode.com/products/mms2r-pdf

## SYNOPSIS

  begin
    require 'mms2r'
  rescue LoadError
    require 'rubygems'
    require 'mms2r'
  end

  # required for the example
  require 'fileutils'

  mail = MMS2R::Media.new(Mail.read('some_saved_mail.file'))
  puts "mail has default carrier subject" if mail.subject.empty?

  # access the sender's phone number
  puts "mail was from phone #{mail.number}"

  # most mail are either image or video, default_media will return the largest
  # (non-advertising) video or image found
  file = mail.default_media
  puts "mail had a media: #{file.inspect}" unless file.nil?

  # finds the largest (non-advertising) text found
  file = mail.default_text
  puts "mail had some text: #{file.inspect}" unless file.nil?

  # mail.media is a hash that is indexed by mime-type.
  # The mime-type key returns an array of filepaths
  # to media that were extract from the mail and
  # are of that type
  mail.media['image/jpeg'].each {|f| puts "#{f}"}
  mail.media['text/plain'].each {|f| puts "#{f}"}

  # print the text (assumes mail had text)
  text = IO.readlines(mail.media['text/plain'].first).join
  puts text

  # save the image (assumes mail had a jpeg)
  FileUtils.cp mail.media['image/jpeg'].first, '/some/where/useful', :verbose => true

  puts "does the mail have quicktime video? #{!mail.media['video/quicktime'].nil?}"

  puts "plus run anything that Mail provides, e.g. #{mail.to.inspect}"

  # check if the mail is from a mobile phone
  puts "mail is from a mobile phone #{mail.is_mobile?}"

  # determine the device type of the phone
  puts "mail is from a mobile phone of type #{mail.device_type?}"

  # inspect default media's exif data if exifr gem is installed and default
  # media is a jpeg or tiff
  puts "mail's default media's exif data is:"
  puts mail.exif.inspect

  # Block support, process and receive all media types of video.
  mail.process do |media_type, files|
    # assumes a Clip model that is an AttachmentFu
    Clip.create(:uploaded_data => files.first, :title => "From phone") if media_type =~ /video/
  end

  # Another AttachmentFu example, Picture model is an AttachmentFu
  picture = Picture.new
  picture.title = mail.subject
  picture.uploaded_data = mail.default_media
  picture.save!

  #remove all the media that was put to temporary disk
  mail.purge

## REQUIREMENTS

* Mail (mail)
* Nokogiri (nokogiri)
* UUIDTools (uuidtools)
* EXIF Reader (exif)

## INSTALL

* sudo gem install mms2r

## SOURCE

git clone git://github.com/monde/mms2r.git

## AUTHORS

Copyright (c) 2007-2012 by Mike Mondragon blog[http://plasti.cx/]

MMS2R's {Flickr page}[http://www.flickr.com/photos/8627919@N05/]

## CONTRIBUTORS

* Luke Francl - blog[http://railspikes.com/]
* Will Jessup - blog[http://www.willjessup.com/]
* Shane Vitarana - blog[http://www.shanesbrain.net/]
* Layton Wedgeworth - company[http://www.beenup2.com/]
* Jason Haruska - blog[http://software.haruska.com/]
* Dave Myron - company[http://contentfree.com/]
* Vijay Yellapragada
* Jesse Dp - {github profile}[http://github.com/jessedp]
* David Alm
* Jeremy Wilkins
* Matt Conway - {github profile}[http://github.com/wr0ngway]
* Kai Kai
* Michael DelGaudio
* Sai Emrys - blog[http://saizai.com]
* Brendan Lim - {github profile}[http://github.com/brendanlim]
* Scott Taylor - {github profile}[http://github.com/smtlaissezfaire]
* Jaap van der Meer - {github profile}[http://github.com/japetheape]
* Karl Baum - {github profile}[http://github.com/kbaum]
* James McGrath - blog[http://jamespmcgrath.com]


History.txt
### 3.9.2 / 2013-09-15 (Abigail Remeltindrinc - record producer)

* 1 minor enhancement
  * psych hack is only for ruby 1.9

### 3.9.1 / 2012-11-15 (#216 - Klokateer afflicted with dwarfism)

* 1 minor enhancement
  * remove rake 0.9.2.2 dependency

### 3.9.0 / 2012-10-18 (Jean-Pierre - A French chef)

* 1 minor enhancement
  * refined character encoding capability with 1.8 rubies and 1.9 rubies.

### 3.8.2 / 2012-08-21 (Seth - Pickles The Drummer's brother)

* 2 minor enhancements
  * Broadened handling of generic "Sent from" footers in smart phones
  * Add ability to process HTC Sensation footer

### 3.8.1 / 2012-07-22 (Calvert - Pickles The Drummer's dad)

* 1 major enhancement
  * Support for Cincinnati Bell - mms.gocbw.com - James McGrath

### 3.8.0 / 2012-07-04 (Dr. Donald Gorfield – Comedy specialist)

* 1 major enhancement
  * Handle MMS from Sprint that have their media attached rather than CDN'd

### 3.7.1 / 2012-06-04 (Abrigail Remeltindtdrinc - The Record Cleaner)

* 2 minor enhancements
  * Improve processing of Sprint media - James McGrath
  * RDoc format documentation - James McGrath

### 3.7.0 / 2012-05-31 (The Teenager - Edgar's brother via Eric's face)

* 2 major enhancements
  * Improve processing of Sprint media - James McGrath
  * Remove Hoe and gemcutter dependencies

### 3.6.4 / 2012-04-22 (Molly - Pickles' mother)

* 1 minor enhancement
  * strip a variation of the iphone default footer

### 3.6.3 / 2012-03-25 (Dethklok Minute Host - host of The Dethklok Minute)

* 1 minor enhancement
  * strip Windows phone default footer

### 3.6.2 / 2012-02-12 (Mr. Gojira - Owner Mr. Gojira's Driving School - retake driver's exam)

* 1 minor enhancement
  * Change user agent given to Sprint so it returns a properly sized image.

### 3.6.1 / 2012-02-11 (Mr. Gojira - Owner Mr. Gojira's Driving School)

* 1 minor enhancement
  * use Net::HTTP#get2 to fetch Sprint content

### 3.6.0 / 2012-01-19 (Agent 216 - assassin, master of infiltration and sabotage)

* 1 major enhancement
  * Ruby 1.8.7, 1.9.2, and 1.9.3 compatible
* 1 bug fix
  * A fix for the utf-8 encoding issue with ruby 1.9.2 - minimul / Christian

### 3.5.1 / 2011-12-11 (Mashed Potato Johnson - oldest living blues guitarist in the world)

* 1 bug fix
  * A fix for the utf-8 encoding issue with ruby 1.9.2 - James McGrath

### 3.5.0 / 2011-12-01 (Dr. Milminiman Lamilim Swimwambli - Marriage expert)

* 1 major bug fix
  * In Ruby 1.9.X MMS2R::Media::Sprint was not fetching content from Sprint's
    CDN correctly because the implementation of Net::HTTP in 1.9.X passes a
    User-Agent header of "Ruby" to which Sprint responds with a fake 500 -
    with help from James McGrath

### 3.4.1 / 2011-11-06 (Fatty Ding-Dongs, Dethklok foster child)

* 2 major enhancements
  * Additional means to detect android, iphone, motorola, nokia, palm handsets
  * Additional known SMS/MMS domains mm.att.net, labwig.net, cdma.sasktel.com

### 3.4.0 / 2011-10-31 (Serveta Skwigelf, Skwisgaar's hot mom)

* 2 major enhancements
  * regexp's in yaml configs are store as a ruby regexp and not a string that
    is evaled.
  * character encoding capability with 1.8 rubies and 1.9 rubies.

### 3.3.1 / 2011-01-01 (Reverend Aslaug Wartooth (deceased), Toki's dad)

* 2 minor enhancements
  * new method #default_html returns the default html attachment
  * #body will return the default plain text, or default stripped html text

### 3.3.0 / 2010-12-31 (Charles Foster Offdensen - CFO, dethklok)

* 1 major enhancement
  * MMS2R::Media::Sprint is now a module that is extended in for processing
    Sprint content rather than being child class of MMS2R::Media
    Extension is the strategy to use for overwriting template methods now

* 1 minor enhancement
  * process new Sprint subject - Jason Hooper
  * new SaskTel mms/sms domain alias sasktel.com
  * new Virgin Mobile mms/sms domain alias yrmobl.com

### 3.2.0 / 2010-08-17 (Antonio "Tony" DiMarco Thunderbottom - bassist)

* 3 minor enhancements
  * updated readme with latest sites known to use mms2r
  * bundler assimilated
  * loading rubygems only if it's required, like rake

### 3.1.0 / 2010-07-09 (Dick "Magic Ears" Knubbler - music producer)

* 8 minor enhancements
  * Detects additional Bell/AT&T domain sbcglobal.net as a handset
  * Detects additional Virgin mobile domains pixmbl.com vmobl.com as a handset
  * Expanded mobile phone detection for handset makers by : Casio, Google Nexus
    One, LG Electronics, Motorola, Pantech, Qualcom, Samsung, Sprint, UTStarCom
  * EXIF Reader (exifr) is integral to MMS2R and is now a required gem
  * upgrade Mail to 2.2.5
  * depend on latest gems
  * corrected example - Jason Hooper
  * added As Seen On The Internet to README that lists sites known to be using
    MMS2R in some fashion

### 3.0.1 / 2010-03-28 (Lavona Succuboso - leader of "Succuboso Explosion")

* 1 minor enhancement
  * support for U.S. Cellular

### 3.0.0 / 2010-02-24 (General Crozier - Chairman of the Joint Chiefs of Staff)

* 1 new feature
  * dependence upon Mail gem rather than TMail gem

### 2.4.1 / 2010-02-07 (Vater Orlaag - political and spiritual specialist)

* 3 minor enhancements
  * make sure filename is free of new lines - laurynas
  * recognize HTC HERO200 as a smart phone
  * recognize HTC Eris as a smart phone

### 2.4.0 / 2009-12-20 (Dr. Nanemiltred Philtendrieden - specialist on celebrity death)

* 1 new feature
  * replace Hpricot with Nokogiri for html parsing of Sprint data

* 3 minor enhancements
  * smartphone identities stored in conf/mms2r_media.yml
  * identify Motorola Droid as a smartphone
  * identify T-Mobile Dash as a smartphone
  * use uuidtools for naming temp directories - kbaum

### 2.3.0 / 2009-08-30 (Snakes 'n' Barrels Greatest Hits)
* 5 new features
  * detect smartphone status/type based on model metadata from jpeg and tiff
    exif data using exifr gem, access exif data with MMS2R::Media#exif

lib/mms2r.rb
#--
# Copyright (c) 2007-2012 by Mike Mondragon (mikemondragon@gmail.com)
#
# Please see the LICENSE file for licensing information.
#++

module MMS2R

  ##
  # A hash of MMS2R processors keyed by MMS carrier domain.

  CARRIERS = {}

  ##
  # Registers a class as a processor for a MMS domain.  Should only be
  # used in special cases such as MMS2R::Media::Sprint for 'pm.sprint.com'

  def self.register(domain, processor_class)
    MMS2R::CARRIERS[domain] = processor_class
  end

  ##
  # A hash of file extensions for common mime-types

  EXT = {
    'text/plain' => 'text',
    'text/plain' => 'txt',
    'text/html' => 'html',
    'image/png' => 'png',
    'image/gif' => 'gif',
    'image/jpeg' => 'jpeg',
    'image/jpeg' => 'jpg',
    'video/quicktime' => 'mov',
    'video/3gpp2' => '3g2'
  }

  class MMS2R::Media

    ##
    # Spoof User-Agent, primarily for the Sprint CDN

    USER_AGENT = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/535.2 (KHTML, like Gecko) Chrome/15.0.874.120 Safari/535.2"
  end

  ##
  # Simple convenience function to make it a one-liner:
  # MMS2R.parse raw_mail or
  # MMS2R.parse File.load(file)
  # MMS2R.parse File.load(path_to_file)
  # Combined w/ the method_missing delegation, this should behave as an enhanced Mail object, more or less.

  def self.parse thing, options = {}
    mail = case
           when File.exist?(thing); Mail.new(open(thing).read)
           when thing.respond_to?(:read); Mail.new(thing.read)
           else
             Mail.new(thing)
           end

    MMS2R::Media.new(mail, options)
  end

  ##
  # Compare original MMS2R results with original mail values and other metrics.
  #
  # Takes a file path, mms2r object, mail object, or mail text blob.

  def self.debug(thing, options = {})
    mms = case thing
          when MMS2R::Media; thing
          when Mail::Message; MMS2R::Media.new(thing, options)
          else
            self.parse(thing, options)
          end

    <<OUT
#{'-' * 80}

  original mail
    #{'from:'.ljust(15)} #{mms.mail.from}
    #{'to:'.ljust(15)} #{mms.mail.to}
    #{'subject:'.ljust(15)} #{mms.mail.subject}

  mms2r
    #{'from:'.ljust(15)} #{mms.from}
    #{'to:'.ljust(15)} #{mms.to}
    #{'subject:'.ljust(15)} #{mms.subject}
    #{'number:'.ljust(15)} #{mms.number}

    default media
      #{mms.default_media.inspect}

    default text
      #{mms.default_text.inspect}
      #{mms.default_text.read if mms.default_text}

    all plain text
      #{(mms.media['text/plain']||[]).each {|t| open(t).read}.join("\n\n")}

    media hash
      #{mms.media.inspect}

OUT
  end

end

%W{ yaml mail fileutils pathname tmpdir yaml digest/sha1 exifr }.each do |g|
  begin
    require g
  rescue LoadError
    require 'rubygems'
    require g
  end
end

if RUBY_VERSION == "1.9" && RUBY_VERSION < "2.0"
  begin
    require 'psych'
    YAML::ENGINE.yamler= 'syck' if defined?(YAML::ENGINE)
  rescue LoadError
  end
end


require File.join(File.dirname(__FILE__), 'ext/mail')
require File.join(File.dirname(__FILE__), 'ext/object')
require File.join(File.dirname(__FILE__), 'mms2r/media')
require File.join(File.dirname(__FILE__), 'mms2r/media/sprint')

MMS2R.register('pm.sprint.com', MMS2R::Media::Sprint)
